flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSource.java

```diff
@@ -108,13 +121,39 @@ public boolean isBounded() {
 		};
 	}
 
-	private InputFormat<RowData, ?> getInputFormat() {
-		// When this table has no partition, just return a empty source.
-		if (!partitionKeys.isEmpty() && getOrFetchPartitions().isEmpty()) {
-			return new CollectionInputFormat<>(new ArrayList<>(), null);
+	private FileSource<RowData> createBulkFormatSource(
+			BulkDecodingFormat<RowData> decodingFormat, ScanContext scanContext) {
+		decodingFormat.applyLimit(pushedDownLimit());
+		decodingFormat.applyFilters(pushedDownFilters());
+		BulkFormat<RowData> bulkFormat = decodingFormat.createRuntimeDecoder(
+				scanContext, getProducedDataType());
+		FileSource.FileSourceBuilder<RowData> builder = FileSource
+				.forBulkFileFormat(bulkFormat, paths());
+		return builder.build();
+	}
+
+	private Path[] paths() {
+		if (partitionKeys.isEmpty()) {
+			return new Path[] {path};
+		} else {
+			return getOrFetchPartitions().stream()
+					.map(FileSystemTableSource.this::toFullLinkedPartSpec)
+					.map(PartitionPathUtils::generatePartitionPath)
+					.map(n -> new Path(path, n))
+					.toArray(Path[]::new);
 		}
+	}
+
+	private long pushedDownLimit() {
+		return limit == null ? Long.MAX_VALUE : limit;
```

Comment:
I think `Long.MAX_VALUE` can't represent no limit, right?

Reply:
We can keep null value

Reply:
We can do something like this:
```
		if (limit != null) {
			decodingFormat.applyLimit(limit);
		}
		if (filters != null && filters.size() > 0) {
			decodingFormat.applyFilters(filters);
		}
```